BootloaderCorePkg/Stage2/Stage2Support.c

```diff
@@ -122,6 +122,79 @@ DisplaySplash (
   return Status;
 }
 
+/**
+  Initialize the virtual console framebuffer back-end.
+
+  @retval EFI_UNSUPPORTED Could not initilaize the console.
+  @retval EFI_SUCCESS     Console initialized.
+**/
+EFI_STATUS
+EFIAPI
+ConsoleInitFrameBuffer (
+  VOID
+  )
+{
+  LOADER_GLOBAL_DATA      *LdrGlobal;
+  FRAME_BUFFER_INFO       *FrameBuffer;
+  FRAME_BUFFER_CONSOLE    *Console;
+  DEBUG_LOG_BUFFER_HEADER *LogBufHdr;
+
+  if (FixedPcdGet32 (PcdConsoleOutDeviceMask) & ConsoleOutFrameBuffer) {
+    LdrGlobal = (LOADER_GLOBAL_DATA *)GetLoaderGlobalDataPointer ();
+    FrameBuffer = (FRAME_BUFFER_INFO *)GetGuidHobData (LdrGlobal->LdrHobList, NULL, &gLoaderFrameBufferInfoGuid);
+    if (FrameBuffer == NULL) {
+      return EFI_UNSUPPORTED;
+    } 
+
+    Console = BuildGuidHob (&gLoaderFrameBufferConsoleGuid, sizeof (FRAME_BUFFER_CONSOLE));
+    InitFrameBufferConsole (FrameBuffer, Console,
+                            FrameBuffer->HorizontalResolution,
+                            FrameBuffer->VerticalResolution,
+                            0, 0);
+    PcdSet64(PcdFrameBufferConsoleBase, (UINT64)(UINTN)Console);
+
+    // If console is outputting serial messages, also print out any messages
+    // stored in log buffer as the console is not initialized until much later.
+    if ((PcdGet32 (PcdConsoleOutDeviceMask) & ConsoleOutSerialPort)
+        && (PcdGet32 (PcdDebugOutputDeviceMask) & DEBUG_OUTPUT_DEVICE_LOG_BUFFER)) {
+      LogBufHdr = (DEBUG_LOG_BUFFER_HEADER *) GetDebugLogBufferPtr ();
+      if (LogBufHdr != NULL) {
+        ConsoleWrite (LogBufHdr->Buffer, LogBufHdr->UsedLength - LogBufHdr->HeaderLength);
+      }
+    }
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Update console framebuffer pointer based on gLoaderFrameBufferInfoGuid HOB.
+
+**/
+VOID
+EFIAPI
+ConsoleUpdateFrameBuffer (
```

Comment:
BuildExtraInfoHob()  in State2Support.c already did this.   I think we can break out one function called  UpdateFrameBuffer().   Then everybody call the same function to update framebuffer info.